src/WakaTimeSendHeartbeatThread.pas

```diff
@@ -0,0 +1,182 @@
+    FCLIPath: string;
+    FAPIKey: string;
+    FFileName: string;
+    FProjectName: string;
+    FIsFile: Boolean;
+    FTotalLines: Integer;
+    FIsWrite: Boolean;
+  protected
+    procedure Execute; override;
+  public
+    constructor Create(CLIPath, APIKey, FileName, ProjectName: string;
+      IsFile: Boolean; TotalLines: Integer; IsWrite: Boolean);
+  end;
+
+implementation
+
+{$I DelphiVersions.inc}
+
+uses
+  Windows,
+  ShellAPI,
+  WakaTimeLogger,
+  StrUtils;
+
+const
+  PluginVersion = '1.73.1'; // Replace with actual plugin version
+
+  {$IFDEF DELPHI_11_3_ALEXANDRIA}
+  DelphiVersion = '10.3';
+  {$ENDIF}
+  {$IFDEF DELPHI_10_4_SYDNEY}
+  DelphiVersion = '10.3';
+  {$ENDIF}
+  {$IFDEF DELPHI_10_3_RIO}
+  DelphiVersion = '10.3';
+  {$ENDIF}
+  {$IFDEF DELPHI_10_2_TOKYO}
+  DelphiVersion = '10.2';
+  {$ENDIF}
+  {$IFDEF DELPHI_10_1_BERLIN}
+  DelphiVersion = '10.1';
+  {$ENDIF}
+  {$IFDEF DELPHI_10_SEATTLE}
+  DelphiVersion = '10.0';
+  {$ENDIF}
+  {$IFDEF DELPHI_XE8}
+  DelphiVersion = 'XE8';
+  {$ENDIF}
+  {$IFDEF DELPHI_XE7}
+  DelphiVersion = 'XE7';
+  {$ENDIF}
+  {$IFDEF DELPHI_XE6}
+  DelphiVersion = 'XE6';
+  {$ENDIF}
+  {$IFDEF DELPHI_XE5}
+  DelphiVersion = 'XE5';
+  {$ENDIF}
+  {$IFDEF DELPHI_XE4}
+  DelphiVersion = 'XE4';
+  {$ENDIF}
+  {$IFDEF DELPHI_XE3}
+  DelphiVersion = 'XE3';
+  {$ENDIF}
+  {$IFDEF DELPHI_XE2}
+  DelphiVersion = 'XE2';
+  {$ENDIF}
+  {$IFDEF DELPHI_XE}
+  DelphiVersion = 'XE';
+  {$ENDIF}
+  {$IFDEF DELPHI_2010}
+  DelphiVersion = '2010';
+  {$ENDIF}
+  {$IFDEF DELPHI_2009}
+  DelphiVersion = '2009';
+  {$ENDIF}
+  {$IFDEF DELPHI_2007_FOR_NET}
+  DelphiVersion = '2007.NET';
+  {$ENDIF}
+  {$IFDEF DELPHI_2007}
+  DelphiVersion = '2007';
+  {$ENDIF}
+  {$IFDEF DELPHI_2006}
+  DelphiVersion = '2006';
+  {$ENDIF}
+  {$IFDEF DELPHI_2005}
+  DelphiVersion = '2005';
+  {$ENDIF}
+  {$IFDEF DELPHI_8_FOR_NET}
+  DelphiVersion = '8.NET';
+  {$ENDIF}
+  {$IFDEF DELPHI_7}
+  DelphiVersion = '7.0';
+  {$ENDIF}
+  {$IFDEF DELPHI_6}
+  DelphiVersion = '6.0';
+  {$ENDIF}
+  UserAgent = 'delphi/' + DelphiVersion + ' delphi-wakatime/' + PluginVersion;
+
+constructor TSendHeartbeatThread.Create(CLIPath, APIKey, FileName,
+  ProjectName: string; IsFile: Boolean; TotalLines: Integer; IsWrite: Boolean);
+begin
+  inherited Create(False);
+  FreeOnTerminate := True;
+  FCLIPath := CLIPath;
+  FAPIKey := APIKey;
+  FFileName := FileName;
+  FProjectName := ChangeFileExt(ProjectName, '');
+  FIsFile := IsFile;
+  FTotalLines := TotalLines;
+  FIsWrite := IsWrite;
+end;
+
+procedure TSendHeartbeatThread.Execute;
+var
+  Commands: TStringList;
+  CommandLine, executable: string;
+  Operation: PChar;
+  FileName: PChar;
+  Parameters: PChar;
+  Directory: PChar;
+  ShowCommand: Integer;
+begin
+  if (FAPIKey = '') then
+  begin
+    TWakaTimeLogger.Log('ApiKey not found =(');
+    exit;
+  end;
+
+  executable := Format('"%swakatime-cli.exe"', [FCLIPath]);
+
+  // Prepare the command line
+  Commands := TStringList.Create;
```

Comment:
What was the main reason to choose to use a StringList and do all this replaces? can you detail what problem have you faced that brings you to this solution? We have to remember that stringlist allocs more memory than just using a string and this can be crucial to memory allocation as the editor events are triggering many times creating threads on background from a variety of actions.

Reply:
Lets see if it impact the usage I will look into just to certify that TStringList is Thread Safe and will make some tests but for now I don't see any problem.  Besides that I don't remember now what was the main reason for me to use the cmd /c to execute the commands but it has one specific reason that I don't remember now :/ 

Lets test it if we find something odd we can roll back the way it was and write some unit tests to avoid it happen again.